base/strings/util.jl

```diff
@@ -388,13 +388,83 @@ _replace(io, repl, str, r, pattern) = print(io, repl)
 _replace(io, repl::Function, str, r, pattern) =
     print(io, repl(SubString(str, first(r), last(r))))
 
-replace(str::String, pat_repl::Pair{Char}; count::Integer=typemax(Int)) =
-    replace(str, equalto(first(pat_repl)) => last(pat_repl); count=count)
-replace(str::String, pat_repl::Pair{<:Union{Tuple{Vararg{Char}},AbstractVector{Char},Set{Char}}};
-        count::Integer=typemax(Int)) =
-    replace(str, occursin(first(pat_repl)) => last(pat_repl), count)
+# replace Char with function that compares with this Char
+pairmap(p::Pair{Char}) = equalto(p.first) => p.second
+
+# replace collection of Char with function that checks occurrence in this collection
+function pairmap(p::Pair{<:Union{Tuple{Vararg{Char}},AbstractVector{Char},Set{Char}}})
+    occursin(p.first) => p.second
+end
+pairmap(p::Pair) = p
+
+# find the first of a list of patterns.
+# if several match at the same start position prefer the longer one
+# if also the length is equal, prefer the first in list
+# return value of the successful search and the pair
+
+function findnextall(pairs::NTuple{N,Pair}, s::AbstractString, st::Integer=1) where N
+    ps = start(pairs)
+    while !done(pairs, ps)
+        p, ps = next(pairs, ps)
+        r = findnext(p.first, s, st)
+        fr = first(r)
+        if fr > 0
+            minstart, minend, minp = fr, last(r), p
+            while !done(pairs, ps)
+                p, ps = next(pairs, ps)
+                r = findnext(p.first, s, st)
```

Comment:
I have to say that I'm skeptical of the whole approach in this PR.  First converting the user inputs to a tuple of `Pair{Function}` and then calling `findnext` on the function seems almost guaranteed to be inefficient, in several ways.

For one thing, storing a bunch of functions in a data structure and then calling `findnext` on them probably makes it impossible for the compiler to specialize `findnext` by inlining the functions.

For another thing, this means that the search procedure is completely re-done for each pair, when I think that a lot of computations could probably be shared.

Reply:
I think, I should disagree :-).
Most of your objections came also to my mind, but I could handle it.
1. the code was copied from the original implementation of `replace`, which has not been removed, so there is no worse performance than before for the case with one pair.
2. the use of `findnext` had been inserted to the code base shortly before this PR. It replaced `searchnext`, which could handle `Char` and collections of `Char` as patterns, which `findmax` does not do. So for those cases, the `equalto` and `occursin` need used. I think, the authors of `findmax` proved, that performance is not worse than for `searchnext`. I could measure a relevant improvement.
3. my function `findnextall` is a slight generalization of `findmax`, which is thought to make simultaneous pattern searches. Given the complexity of `findnext` it was thought better to re-use that than to re-implement the multiple find next from scratch. The rest of the code of `replace` was re-used, replacing `findnext`with `findnextall`.
4. I see an improvement possibility, by extending the `findnext` function by a new optional argument, which can prevent to search the full input string, when already a bound for the starting point of the match is known. That could decrease search times by up to 50%.